// @wry/trie/rollup.config.js
import typescriptPlugin from 'rollup-plugin-typescript2';
import typescript from 'typescript';

const globals = {
  __proto__: null,
  tslib: "tslib",
};

function external(id) {
  return id in globals;
}

export default [{
<<<<<<<< HEAD:server/node_modules/ts-invariant/rollup.config.js
  input: "src/invariant.ts",
  external,
  output: {
    file: "lib/invariant.esm.js",
========
  input: "src/equality.ts",
  external,
  output: {
    file: "lib/equality.esm.js",
>>>>>>>> develop:client/src/node_modules/@wry/trie/rollup.config.js
    format: "esm",
    sourcemap: true,
    globals,
  },
  plugins: [
    typescriptPlugin({
      typescript,
      tsconfig: "./tsconfig.rollup.json",
    }),
  ],
}, {
<<<<<<<< HEAD:server/node_modules/ts-invariant/rollup.config.js
  input: "lib/invariant.esm.js",
  external,
  output: {
    // Intentionally overwrite the invariant.js file written by tsc:
    file: "lib/invariant.js",
    format: "cjs",
    exports: "named",
    sourcemap: true,
    name: "ts-invariant",
========
  input: "lib/equality.esm.js",
  external,
  output: {
    // Intentionally overwrite the equality.js file written by tsc:
    file: "lib/equality.js",
    format: "cjs",
    exports: "named",
    sourcemap: true,
    name: "equality",
>>>>>>>> develop:client/src/node_modules/@wry/trie/rollup.config.js
    globals,
  },
}];
